File: src/dp/convert_to_letter_string.java
package dp;

public class convert_to_letter_string {
  public static int number(String str) {
    if (str == null || str.length() == 0)
      return 0;

    return process(str.toCharArray(), 0);
  }

  private static int process(char[] str, int i) {
    if (i == str.length)
      return 1;
    if (str[i] == '0')
      return 0;

    int ways = process(str, i + 1);
    if (i + 1 < str.length && (str[i] - '0') * 10 + (str[i + 1] - '0') < 27) {
      ways += process(str, i + 2);
    }
    return ways;
  }

  public static int dp(String s) {
    if (s == null || s.length() == 0)
      return 0;
    char[] str = s.toCharArray();
    int N = str.length;
    int[] dp = new int[N + 1];
    dp[N] = 1;
    for (int i = N - 1; i >= 0; i--) {
      if (str[i] != '0') {
        int ways = dp[i + 1];
        if (i + 1 < str.length && (str[i] - '0') * 10 + (str[i + 1] - '0') < 27) {
          ways += dp[i + 2];
        }
        dp[i] = ways;
      }
    }
    return dp[0];
  }

  // 为了测试
  public static String randomString(int len) {
    char[] str = new char[len];
    for (int i = 0; i < len; i++) {
      str[i] = (char) ((int) (Math.random() * 10) + '0');
    }
    return String.valueOf(str);
  }

  public static void main(String[] args) {
    int N = 30;
    int testTime = 1000000;
    System.out.println("测试开始");
    for (int i = 0; i < testTime; i++) {
      int len = (int) (Math.random() * N);
      String s = randomString(len);
      int ans0 = number(s);
      int ans1 = dp(s);
      // int ans2 = dp2(s);
      if (ans0 != ans1) {
        System.out.println(s);
        System.out.println(ans0);
        System.out.println(ans1);
        // System.out.println(ans2);
        System.out.println("Oops!");
        break;
      }
    }
    System.out.println("测试结束");
  }
}

File: src/dp/convert_to_letter_string.cpp
#include <iostream>

using namespace std;

class Solution {
 private:
  int process(string &str, int i) {
    if (i == str.length()) return 1;
    if (str[i] == '0') return 0;

    int ways = process(str, i + 1);
    if (i + 1 < str.length() && (str[i] - '0') * 10 + str[i + 1] - '0' < 27) {
      ways += process(str, i + 2);
    }
    return ways;
  }

 public:
  int number(string &str) {
    if (str.empty() || str.length() == 0) return 0;
    return process(str, 0);
  }

  int dp(string &str) {
    if (str.empty() || str.length() == 0) return 0;
    int N = str.length();
    vector<int> dp(N + 1, 0);
    dp[N] = 1;
    for (int i = N - 1; i >= 0; i--) {
      if (str[i] != '0') {
        int ways = dp[i + 1];
        if (i + 1 < str.length() &&
            (str[i] - '0') * 10 + str[i + 1] - '0' < 27) {
          ways += dp[i + 2];
        }
        dp[i] = ways;
      }
    }
    return dp[0];
  }

  // TEST
  string randomString(int len) {
    vector<char> str(len, 0);
    for (int i = 0; i < len; i++) {
      str[i] = (char)((int)(rand() % 10) + '0');
    }
    string res = "";
    for (int i = 0; i < len; i++) {
      res += str[i];
    }

    return res;
  }
};

int main() {
  int N = 30;
  int testTime = 1000000;
  Solution s;

  cout << "测试开始" << endl;
  for (int i = 0; i < testTime; i++) {
    int len = (int)(rand() % N);
    string str = s.randomString(len);
    int ans0 = s.number(str);
    int ans1 = s.dp(str);
    if (ans0 != ans1) {
      cout << str << endl;
      cout << ans0 << endl;
      cout << ans1 << endl;
      cout << "Oops!" << endl;
      break;
    }
  }

  cout << "测试结束" << endl;
  return 0;
}